dali/core/geom_mat_test.cu

```diff
@@ -0,0 +1,344 @@
+    { 5, 6, 7, 8 },
+    { 9, 10, 11, 12 },
+  }};
+  mat4x3 transposed = m.T();
+  for (size_t i = 0; i < m.rows; i++)
+    for (size_t j = 0; j < m.cols; j++)
+      EXPECT_EQ(transposed(j, i), m(i, j));
+}
+
+TEST(Mat, Mul) {
+  mat3x4 m1 = {{
+    { 1, 10, 100, 1000 },
+    { 2, 20, 200, 2000 },
+    { 4, 40, 400, 4000 },
+  }};
+  mat<4, 2> m2 = {{
+    { 3, 1 },
+    { 5, 2 },
+    { 7, 3 },
+    { 9, 4 }
+  }};
+  mat3x2 ref = {{
+    { 3+50+700+9000,     1+20+300+4000 },
+    { 6+100+1400+18000,  2+40+600+8000 },
+    { 12+200+2800+36000, 4+80+1200+16000 }
+  }};
+  mat3x2 result = m1*m2;
+  for (size_t i = 0; i < 3; i++)
+    for (size_t j = 0; j < 2; j++)
+      EXPECT_EQ(result(i, j), ref(i, j)) << "@ " << i << ", " << j;
+}
+
+TEST(Mat, Mul4x4) {
+  mat4 m1 = {{
+    { 1, 10, 100, 1000 },
+    { 2, 20, 200, 2000 },
+    { 3, 30, 300, 3000 },
+    { 4, 40, 400, 4000 },
+  }};
+  mat4 m2 = {{
+    {  1,  2,  3,  4 },
+    {  5,  6,  7,  8 },
+    {  9, 10, 11, 12 },
+    { 13, 14, 15, 16 },
+  }};
+
+  mat4 ref = {{
+    { 13951, 15062, 16173, 17284 },
+    { 27902, 30124, 32346, 34568 },
+    { 41853, 45186, 48519, 51852 },
+    { 55804, 60248, 64692, 69136 }
+  }};
+  mat4 result = m1 * m2;
+  for (size_t i = 0; i < 4; i++)
+    for (size_t j = 0; j < 4; j++)
+      EXPECT_EQ(result(i, j), ref(i, j)) << "@ " << i << ", " << j;
+}
+
+
+TEST(Mat, CompoundAssignScalar) {
+  #define TEST_OP(op)\
+  {\
+    m1 op##= s;\
+    for (size_t i = 0; i < 3; i++)\
+      for (size_t j = 0; j < 4; j++)\
+        EXPECT_EQ(m1(i, j), orig(i, j) op s);\
+    m1 = orig;\
+  }
+
+  {
+    mat3x4 m1 = {{
+      { 1, 10, 100, 1000 },
+      { 2, 20, 200, 2000 },
+      { 4, 40, 400, 4000 },
+    }};
+    float s = 42;
+    auto orig = m1;
+    TEST_OP(+)
+    TEST_OP(-)
+    TEST_OP(*)
+    TEST_OP(/)
+  }
+
+  {
+    imat3x4 m1 = {{
+      { 1, 10, 100, 1000 },
+      { 2, 20, 200, 2000 },
+      { 4, 40, 400, 4000 },
+    }};
+    int s = 0b1010101;
+    auto orig = m1;
+
+    TEST_OP(&)
+    TEST_OP(|)
+    TEST_OP(^)
+  }
+}
+
+struct cuda_rng {
+  DALI_HOST_DEV int operator()() {
+    return seed = (seed * 1103515245 + 12345) & 0x7fffffff;
+  }
+  unsigned seed;
+};
+
+template <size_t rows, size_t cols, typename T, typename RNG>
+DALI_HOST_DEV
+void RandomFill(mat<rows, cols, T> &m, RNG &rng, int lo, int hi) {
+  for (size_t i = 0; i < rows; i++)
+    for (size_t j = 0; j < cols; j++)
+      m(i, j) = rng()%(hi-lo) + lo;
+}
+
+TEST(Mat, BinOp) {
+#define TEST_OPERATOR_SIZE(op, mat_type) { \
+    mat_type m1, m2; \
+    int lo = 0, hi = 31; \
+    cuda_rng rng = { 12345 }; \
+    RandomFill(m1, rng, lo, hi); RandomFill(m2, rng, lo, hi); \
+    mat_type result = m1 op m2; \
+    for (size_t i = 0; i < result.rows; i++) \
+      for (size_t j = 0; j < result.cols; j++) \
+        EXPECT_EQ(result(i, j), m1(i, j) op m2(i, j)); \
+    int scalar = rng()&31; \
+    result = m1 op scalar; \
+    for (size_t i = 0; i < result.rows; i++) \
```

Comment:
Doesn't that cover `TEST_OP` as well?

Reply:
Actually, test_op tests compound assignment. I should rename it.